### ISSUE TITLE
tsserver never ends loading intellisense

### ISSUE THREAD

**Issue by aihornmac** [microsoft/TypeScript#30618]
**TypeScript Version:**  3.4.0-rc also 3.3.3

**Search Terms:**
tsserver stuck loading intellisense

**Code**

[generic-data-access-layer.1.zip](https://github.com/Microsoft/TypeScript/files/3014569/generic-data-access-layer.1.zip)
[tsserver.log](https://github.com/Microsoft/TypeScript/files/3014571/tsserver.log)
![image](https://user-images.githubusercontent.com/7410810/55099259-5c2cd980-50fa-11e9-8fb9-f6d2d56a7a50.png)

**Expected behavior:**
Loaded in short time

**Actual behavior:**
Never success

**Playground Link:** 

**Related Issues:**

**Comment by aihornmac** [comment]
And also compiling with `tsc` results in out of memory, already tried `node --max-old-space-size=4096 ../../node_modules/.bin/tsc --pretty -p src/tsconfig.json`

The code base is small enough, I reckon that normally it shouldn't cause performance problem.

**Comment by rbuckton** [comment]
The code base is small, but generates some fairly large and complex types. I ran your repro through several different versions of TypeScript. It may have been "fixed" in 3.7, as that version no longer crashes, and remained so through 4.0. However, as of 4.1.0-rc we report the following error due to the complexity of the mapped type that is generated:

```
$ npx typescript@rc -p src/tsconfig.json
npx: installed 1 in 4.112s
src/model/element.ts:5:14 - error TS7056: The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.

5 export const Element = union(Text, Image)
               ~~~~~~~

src/model/layout.ts:16:14 - error TS7056: The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.

16 export const LayoutRoot = model('layout root')
                ~~~~~~~~~~

src/model/layout.ts:21:14 - error TS7056: The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.

21 export const LayoutNode = model('layout node')
                ~~~~~~~~~~


Found 3 errors.
```

I've attached a slightly modified version of your repro with a `package.json` that contains a minimal set of dependencies and an updated `tsconfig.json` that I inferred from the `tsserver.log` you posted:

- [repro.zip](https://github.com/microsoft/TypeScript/files/5527577/repro.zip)

**Comment by rbuckton** [comment]
I'm going to tentatively mark this as "fixed", but will revisit pending some additional discussion with the team.
